components/builder-depot/src/server.rs

```diff
@@ -840,24 +858,112 @@ fn list_packages(req: &mut Request) -> IronResult<Response> {
 }
 
 fn list_channels(req: &mut Request) -> IronResult<Response> {
-    let depot = req.get::<persistent::Read<Depot>>().unwrap();
-    let channels = try!(depot.datastore.channels.all());
-    let body = serde_json::to_string(&channels).unwrap();
+    let lock = req.get::<persistent::State<Depot>>().unwrap();
+    let mut depot = lock.write().expect("depot read lock is poisoned");
+    let params = req.extensions.get::<Router>().unwrap();
 
-    let mut response = Response::with((status::Ok, body));
-    dont_cache_response(&mut response);
-    Ok(response)
+    if let Some(origin) = params.find("origin") {
+        let channels = depot.datastore.channels.all(origin);
+        let body = serde_json::to_string(&channels).unwrap();
+        let mut response = Response::with((status::Ok, body));
+        dont_cache_response(&mut response);
+        Ok(response)
+    } else {
+        Ok(Response::with(status::BadRequest))
+    }
+}
+
+fn create_channel(req: &mut Request) -> IronResult<Response> {
+    let lock = req.get::<persistent::State<Depot>>().unwrap();
+    let mut depot = lock.write().expect("depot write lock is poisoned");
+    let params = req.extensions.get::<Router>().unwrap();
+
+    let origin = match params.find("origin") {
+        Some(origin) => origin.to_string(),
+        _ => return Ok(Response::with(status::BadRequest)),
+    };
+
+    let channel = match params.find("channel") {
+        Some(channel) => channel.to_string(),
+        _ => return Ok(Response::with(status::BadRequest)),
+    };
+
+    // let's make sure this origin actually exists
+    let mut conn = Broker::connect().unwrap();
+    if try!(get_origin(&mut conn, &origin)).is_none() {
+        return Ok(Response::with(status::NotFound));
+    }
+
+    match depot.datastore.channels.create(origin, channel) {
+        Ok(_) => Ok(Response::with(status::Created)),
+        Err(Error::ChannelAlreadyExists(_)) => Ok(Response::with(status::NotModified)),
```

Comment:
A 409 "conflict" is more appropriate here. NotModified should be used for cases where a PUT doesn't update a resource